logutil/pretty_handler.go

```diff
@@ -0,0 +1,330 @@
+package logutil
+
+import (
+	"bytes"
+	"context"
+	"encoding"
+	"fmt"
+	"io"
+	"log/slog"
+	"reflect"
+	"slices"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/TouchBistro/goutils/color"
+)
+
+// PrettyHandler is a Handler that writes Records to an io.Writer in a pretty format that looks like so:
+//
+// DEBUG some log message foo=bar
+type PrettyHandler struct {
+	opts        PrettyHandlerOptions
+	w           io.Writer
+	mu          sync.Mutex
+	c           color.Colorer
+	attrsList   []attrsNode
+	groupPrefix string
+	groups      []string
+}
+
+// PrettyHandlerOptions are options for a PrettyHandler.
+// A zero value consists entirely of default values.
+type PrettyHandlerOptions struct {
+	// AddSource adds source code position information to the log using
+	// the SourceKey attribute.
+	AddSource bool
+
+	// Level reports the minimum record level that will be logged.
+	// See the Level field of [slog.HandlerOptions].
+	Level slog.Leveler
+
+	// ReplaceAttr is called to rewrite each non-group attribute before it is logged.
+	// See the ReplaceAttr field of [slog.HandlerOptions].
+	ReplaceAttr func(groups []string, a slog.Attr) slog.Attr
+
+	// ForceQuote forces quoting of all values.
+	// By default, quoting will only be applied if required.
+	ForceQuote bool
+
+	// Disables using colours in logs.
+	DisableColor bool
+}
+
+// NewPrettyHandler creates a new PrettyHandler that writes to the given writer,
+// using the given options. If opts is nil, the default options are used.
+func NewPrettyHandler(w io.Writer, opts *PrettyHandlerOptions) *PrettyHandler {
+	var o PrettyHandlerOptions
+	if opts != nil {
+		o = *opts
+	}
+	if o.Level == nil {
+		o.Level = slog.LevelInfo
+	}
+	var c color.Colorer
+	c.SetEnabled(!o.DisableColor)
+	return &PrettyHandler{opts: o, w: w, c: c}
+}
+
+func (h *PrettyHandler) clone() *PrettyHandler {
+	return &PrettyHandler{
+		opts:        h.opts,
+		w:           h.w,
+		c:           h.c,
+		attrsList:   slices.Clip(h.attrsList),
+		groupPrefix: h.groupPrefix,
+		groups:      slices.Clip(h.groups),
+	}
+}
+
+func (h *PrettyHandler) Enabled(_ context.Context, level slog.Level) bool {
+	return level >= h.opts.Level.Level()
+}
+
+func (h *PrettyHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	if len(attrs) == 0 {
+		return h
+	}
+	h2 := h.clone()
+	h2.attrsList = append(h2.attrsList, attrsNode{groupPrefix: h2.groupPrefix, groups: h.groups, attrs: attrs})
```

Comment:
The way it is rn it will allow duplicates (i.e. doesn't handle them). However this matches the behaviour of the `slog.TextHandler`. I modelled `PrettyHandler` after `slog.TextHandler`, basically I want it to have similar semantics but format stuff in a nice way for the terminal.
![image](https://github.com/TouchBistro/goutils/assets/22629536/75cfda82-0955-47e1-803a-5b1154b1af15)

Open to changing this if you want, but this was a conscious decision I made given the behaviour of `slog`.

Reply:
Yeah fair if that's what slog does. Seems weird to me (because i'm used to these being maps, so you can't have dupes) but we can be consistent with it

Reply:
There's actually an open issue about this for `JSONHandler`. https://github.com/golang/go/issues/59365

We can easily fix this in the future if we find it's a problem. I don't expect this to come up really in practice though